### ISSUE TITLE
`paddle.unstack` followed by `paddle.grad` crashes the process on CPU

### ISSUE THREAD

**Issue by rookieLiu2018** [PaddlePaddle/Paddle#79188]
### bug描述 Describe the Bug

paddle.unstack works on a valid CPU tensor, but computing gradients through one of the unstacked outputs causes the Python process to crash instead of returning a gradient or raising a Python exception.

```python
import paddle

paddle.disable_static()
paddle.set_device("cpu")


def fn(x):
    x.stop_gradient = False
    parts = paddle.unstack(x, axis=0)
    y = parts[1]
    loss = paddle.sum(y * y)
    grad = paddle.grad(loss, x)[0]
    return loss, grad


x = paddle.arange(12, dtype="float32").reshape([3, 4]) / 10.0

print("paddle:", paddle.__version__, flush=True)
print("about to run paddle.unstack + paddle.grad on legal CPU input", flush=True)

loss, grad = fn(x)

print("loss:", loss.numpy().tolist(), flush=True)
print("grad:", grad.numpy().tolist(), flush=True)
print("expected process exit: 0", flush=True)    #crash
```
The process should not crash. It should return the correct gradient or raise a normal Python exception.

### 其他补充信息 Additional Supplementary Information

`Paddle Version: 3.3.0`
